python/beeai_framework/workflows/workflow.py

```diff
@@ -116,39 +144,72 @@ def set_start(self, name: K) -> "Workflow[T, K]":
-                next = run.steps[-1].name
-            elif step_next is None or step_next == Workflow.NEXT:
-                next = self._find_step(next).next or Workflow.END
-            else:
-                next = step_next
-
-        return run
+    def run(self, state: ModelLike[T], options: ModelLike[WorkflowRunOptions] | None = None) -> Run[WorkflowRun[T, K]]:
+        options = to_model_optional(WorkflowRunOptions, options)
+
+        async def run_workflow(context: RunContext) -> Awaitable[WorkflowRun[T, K]]:
+            run = WorkflowRun[T, K](state=to_model(self._schema, state))
+            # handlers = WorkflowRunContext(steps=run.steps, signal=context.signal, abort=lambda r: context.abort(r))
+            next = self._find_step(self.start_step or self.step_names[0]).current or Workflow.END
+
+            while next and next != Workflow.END:
+                step = self.steps.get(next)
+                if step is None:
+                    raise WorkflowError(f"Step '{next}' was not found.")
+
+                await context.emitter.emit("start", {"run": run, "step": next})
+
+                try:
+                    step_res = WorkflowStepRes[T, K](name=next, state=run.state.model_copy(deep=True))
+                    run.steps.append(step_res)
+
+                    if inspect.iscoroutinefunction(step.handler):
+                        step_next = await step.handler(step_res.state)  # , handlers)
+                    else:
+                        step_next = await asyncio.to_thread(step.handler, step_res.state)  # handlers)
+
+                    check_model(step_res.state)
+                    run.state = step_res.state
+
+                    # Route to next step
+                    if step_next == Workflow.START:
+                        next = run.steps[0].name
+                    elif step_next == Workflow.PREV:
+                        next = run.steps[-2].name
+                    elif step_next == Workflow.SELF:
+                        next = run.steps[-1].name
+                    elif step_next is None or step_next == Workflow.NEXT:
+                        next = self._find_step(next).next or Workflow.END
+                    else:
+                        next = step_next
+
+                    await context.emitter.emit(
+                        "success",
+                        {
+                            "run": run.model_copy(),
+                            "state": run.state,
+                            "step": step,
+                            "next": next,
+                        },
+                    )
+                except Exception as err:
+                    await context.emitter.emit(
+                        "error",
+                        {
+                            "run": run.model_copy(),
+                            "step": next,
+                            "error": err,
+                        },
+                    )
+                    raise err
```

Comment:
Error must always be instanceof `FrameworkError`. Use `FrameworkError.ensure(...)`

Reply:
and add it to the event data

Reply:
https://github.com/i-am-bee/beeai-framework/pull/416/commits/6555bd7cdd464b366304d0ad40ecaf180012f78d